### PR TITLE
[Playground] GetMetadata() endpoint

### LINKED ISSUE #21925: [Playground Task]: Sort out code generation for constants in frontend
### What needs to happen?

When building, the frontend uses constants with backend URLs from `/playground/frontend/lib/config.g.dart`.

Looks like this file is generated from `/playground/frontend/gradle.properties` in `/playground/frontend/build.gradle`. But `gradle.properties` is behind `config.g.dart` which has been edited manually. It means that if code generation is re-ran, the changes that are now in production will be lost.

Moreover, it looks like `gradle.properties` is in turn generated in `/playground/terraform/build.gradle.kts`. And `build.gradle.kts` is in turn behind `gradle.properties` which has been edited manually to change at least one constant: `analyticsUA`. It means that if this code generation is re-ran, these changes will be lost too.

This should be sorted out, the differences should be fixed, and some docs on the generation pipeline should be provided.

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### LINKED ISSUE #22271: [Task]: Display Playground front-end version
### What needs to happen?

We need the playground front-end version to show in the UI to check if a new version is deployed alright. Anything will do: build number, build timestamp etc.

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### LINKED ISSUE #22542: [Task]: Display Beam version in Playground
### What needs to happen?

Display Beam version in Playground

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### LINKED ISSUE #24200: [Task]: Version-control the default config.g.dart
### What needs to happen?

`/playground/frontend/lib/config.g.dart` is required to build both Playground and Tour of Beam Flutter apps. It is generated during the new GKE deployment but for development purposes we must have this file in the repository so the project is immediately runnable after checkout.

However due to infrastructure migration issues we temporarily cannot have this file in the repository, and this must be fixed.

Meanwhile the workaround is copy an example file after checkout:

```bash
cp playground/frontend/lib/config.example.dart playground/frontend/lib/config.g.dart
```

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### PR DESCRIPTION
**Pull request by TSultanov**
# Backend
Add new endpoint to the Playground runners which returns metadata. Returned metadata includes:
- Type of Beam SDK (Java, Go, Python, Scio)
- Version of Beam SDK
- Commit from which the runner was built
- Timestamp of the commit

# Frontend
- Display frontend version
- Display backend version information

resolves #22271
resolves #22542
resolves #21925
resolves #24200

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/api/v1/api.proto [hunk 1]
```diff
@@ -314,6 +314,17 @@ message GetSnippetResponse {
...
 }
 
+// GetMetadataRequest represents request for runner metadata
+message GetMetadataRequest {}
+
+// GetMetadataResponse contains metadata about the runner
+message GetMetadataResponse {
+  string runner_sdk = 1;
+  string build_commit_hash = 2;
+  string build_commit_timestamp = 3;
```
Comment (alexeyinkin) [inline playground/api/v1/api.proto:324]:
```suggestion
  int64 build_commit_seconds_since_epoch = 3;
```
?

In Dart core these are called `millisecondsSinceEpoch`, `microsecondsSinceEpoch`, etc. to avoid confusion since people nowadays expect varying units of timestamp.

File: playground/backend/containers/git-functions.gradle [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ext.getGitCommitHash = () -> {
+   def stdout = new ByteArrayOutputStream()
+   exec {
+      executable 'git'
```
Comment (alexeyinkin) [inline playground/backend/containers/git-functions.gradle:22]: I think we should use Kotlin for new endeavors. Or at least simplify possible migration in the future with
```suggestion
      executable("git")
```

File: playground/backend/containers/java/build.gradle [hunk 3]
```diff
@@ -71,7 +72,9 @@ docker {
   tags containerImageTags()
   buildArgs(['BEAM_VERSION': project.rootProject.hasProperty(["sdk-tag"]) ?
                            project.rootProject["sdk-tag"] :
-                           default_beam_version ])
+                           default_beam_version,
+             'GIT_COMMIT': getGitCommitHash(),
+             'GIT_TIMESTAMP': getGitCommitTimestamp()] )
```
Comment (alexeyinkin) [inline playground/backend/containers/java/build.gradle:77]: Are there auto-formatters or at least guides? This looks random. Maybe at least add a trailing comma?
```suggestion
             'GIT_TIMESTAMP': getGitCommitTimestamp(),
  ])
```

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: playground/backend/containers/git-functions.gradle (hunk 2 above)
Comment (TSultanov) [reply playground/backend/containers/git-functions.gradle]: I haven't found any good ways how to import functions defined in Kotlin DSL into Gradle Groovy. If you have suggestions, I'll be glad to hear them

File: playground/frontend/playground_components/tools/generate_build_metadata.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+# Run this from the playground_components project root
+FILE='lib/src/build_metadata.g.dart'
```
Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:19]: @alexeyinkin  Can you also add this file to `.gitignore` so if user would build the project it wouldn't pollute `git status`?

File: playground/frontend/playground_components/tools/generate_build_metadata.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+echo '// GENERATED CODE - DO NOT MODIFY BY HAND' > $FILE
```
Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:21]: Use the following pattern to put large snippets of text into files:
```shell
cat > $FILE << EOF
....
EOF
```

File: playground/frontend/Dockerfile [hunk 6]
```diff
@@ -23,8 +23,13 @@ ARG FLUTTER_VERSION=3.7.3
 # Setting up.
 # ------------------------------------------------
 
+COPY . /app
+WORKDIR /app
+
 RUN apt-get update && apt-get install -y wget xz-utils git
 
+RUN cat /app/playground_components/lib/src/build_metadata.g.dart
```
Comment (TSultanov) [inline playground/frontend/Dockerfile:31]: Remove

File: playground/frontend/lib/config.g.dart [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * limitations under the License.
+ */
+
+// This file was auto-generated and used to contain backend URLs.
+// They are now determined automatically
+// as of https://github.com/apache/beam/pull/25610
+// TODO(alexeyinkin): When it is merged, find a better place for the remaining
+//  constant and delete this file.
+
+const String kAnalyticsUA = 'G-BXFP2FNCKC';
```
Comment (TSultanov) [inline playground/frontend/lib/config.g.dart:25]: Rename to `config.dart` or similar as it no longer is a generated file

File: playground/frontend/playground_components/tools/generate_build_metadata.sh (hunk 4 above)
Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:19]: Also, can we determine the path to the script location and use it to construct the full path to this generated file to drop the requirement of being in the playground_components project root?
Or, pass this value as an argument from a gradle task
Or, why not generate this file from gradle directly?

Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh:19]: We have this file in the repository so that the project is immediately runnable after checkout. So adding it to .gitignore has no effect. Unfortunately, it will always pollute `git status`.

An alternative would be to write these data to a git-ignored .txt file and to download when the frontend starts. The cost is +1 request on each page load. I would rather not do it.

File: playground/frontend/lib/config.g.dart (hunk 7 above)
Comment (alexeyinkin) [reply playground/frontend/lib/config.g.dart:25]: This is to prevent merging conflicts because there are other PRs to change this file. When analytics lands, I will.

File: playground/frontend/playground_components/lib/src/repositories/code_repository.dart [hunk 8]
```diff
@@ -35,9 +35,11 @@ const kProcessingStartedText = 'The processing has started\n';
 
 // TODO(alexeyinkin): Rename. This is not a repository but a higher level client.
 class CodeRepository {
-  final CodeClient _client;
+  final CodeClient client;
```
Comment (Malarg) [inline playground/frontend/playground_components/lib/src/repositories/code_repository.dart:38]: Client changed to public here just because of one method in `build_metadata.dart`. I suggest to keep this field incapsulated and turn back to private

**Review by Malarg — CHANGES_REQUESTED** [review]

File: playground/frontend/playground_components/tools/generate_build_metadata.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+echo '// This file is generated during deployment to contain data about the commit.' >> $FILE
+echo '// The copy of this file stored in the repository is for development purpose' >> $FILE
+echo '// so the project can be started locally.' >> $FILE
+echo '// It is safe to commit changes here everytime you run code generation.' >> $FILE
+
+echo -n 'const buildCommitHash = ' >> $FILE
+git rev-parse --sq HEAD >> $FILE
+echo ';' >> $FILE
+
+echo -n 'const buildCommitSecondsSinceEpoch = ' >> $FILE
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh]: `+`

File: playground/frontend/Dockerfile (hunk 6 above)
Comment (alexeyinkin) [reply playground/frontend/Dockerfile:31]: I would rather keep it in build logs. If anything goes wrong with generation, it will be problematic to extract this file content.

File: playground/frontend/playground_components/tools/generate_build_metadata.sh (hunk 4 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh]: `+` determining from the script location.
I would rather not use Gradle for this because Groovy files are currently messy and we cannot do this in Kotlin.

File: playground/frontend/playground_components/lib/src/build_metadata.g.dart [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// GENERATED CODE - DO NOT MODIFY BY HAND
+//
+// This file is generated during deployment to contain data about the commit.
+// The copy of this file stored in the repository is for development purpose
+// so the project can be started locally.
+// It is safe to commit changes here everytime you run code generation.
+const buildCommitHash = '0d41f157812fdc3227a6ea93429d3096aa03a850' ;
```
Comment (TSultanov) [inline playground/frontend/playground_components/lib/src/build_metadata.g.dart:7]: I don't like that we commit these values into the repo. Can we write something like `unknown` here?

File: playground/frontend/playground_components/tools/generate_build_metadata.sh [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+cat > $FILE << EOF
+// GENERATED CODE - DO NOT MODIFY BY HAND
+//
+// This file is generated during deployment to contain data about the commit.
+// The copy of this file stored in the repository is for development purpose
+// so the project can be started locally.
+// It is safe to commit changes here everytime you run code generation.
+EOF
+
+printf 'const buildCommitHash = ' >> $FILE
```
Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:30]:
```suggestion
printf 'const buildCommitHash = %s;\n' $(git rev-parse --sq HEAD)
```

@alexeyinkin

File: playground/frontend/playground_components/tools/generate_build_metadata.sh [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+git rev-parse --sq HEAD >> $FILE
+echo ';' >> $FILE
+
+printf 'const buildCommitSecondsSinceEpoch = ' >> $FILE
```
Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:34]:
```suggestion
printf 'const buildCommitSecondsSinceEpoch = %s;\n' $(git show -s --format=%ct HEAD) >> $FILE
```

Comment (TSultanov) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh:34]: Or better
```bash
GIT_COMMIT_TIMESTAMP=$(git show -s --format=%ct HEAD)
...
printf 'const buildCommitSecondsSinceEpoch = %s;\n' $GIT_COMMIT_TIMESTAMP >> $FILE
```

File: playground/frontend/playground_components/tools/generate_build_metadata.sh [hunk 13]
```diff
@@ -0,0 +1,39 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
```
Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:1]: @alexeyinkin  add
```suggestion
#!/bin/bash
#  Licensed to the Apache Software Foundation (ASF) under one
```

Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:1]: @alexeyinkin also, please run `shellcheck` linter tool on this script

File: playground/frontend/playground_components/lib/src/build_metadata.g.dart (hunk 10 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/build_metadata.g.dart]: `+`

File: playground/frontend/playground_components/tools/generate_build_metadata.sh (hunk 11 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh]: `+`

File: playground/frontend/playground_components/tools/generate_build_metadata.sh (hunk 12 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh]: `+`

File: playground/frontend/playground_components/tools/generate_build_metadata.sh (hunk 13 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh:2]: `+`

Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh:2]: `+`

File: playground/frontend/Dockerfile (hunk 6 above)
Comment (alexeyinkin) [reply playground/frontend/Dockerfile]: Deleted.

File: playground/frontend/playground_components/tools/generate_build_metadata.sh [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+cat > "$FILE" << EOF
+// GENERATED CODE - DO NOT MODIFY BY HAND
+//
+// This file is generated during deployment to contain data about the commit.
+// The copy of this file stored in the repository is for development purpose
+// so the project can be started locally.
+// Keep these constants \`null\` in the repository.
+EOF
+
+echo "const buildCommitHash = '$1';" >> "$FILE"
```
Comment (TSultanov) [inline playground/frontend/playground_components/tools/generate_build_metadata.sh:33]: Wonder if moving this into the `cat > .. << EOF` block will work. If so, this will be a nice looking script

File: playground/frontend/playground_components/lib/src/controllers/build_metadata.dart [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+/// Obtains versions from the backend.
+class BuildMetadataController extends ChangeNotifier {
+  ComponentVersion? _routerVersion;
+  Future<GetMetadataResponse>? _routerVersionFuture;
+
+  final _runnerVersions = <Sdk, ComponentVersion>{};
+  final _runnerVersionFutures = <Sdk, Future<GetMetadataResponse>>{};
+
+  /// Returns the router version and starts loading if it is not started yet.
+  ComponentVersion? get routerVersion {
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/controllers/build_metadata.dart]: `+`

File: playground/frontend/playground_components/lib/src/repositories/backend_urls.dart [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+/// with getMetadata() call. This results in slower initialization for
+/// custom stages but keeps the configuration simple.
+Future<Uri> _getBackendUrl(String node) async {
+  final urls = _getBackendUrlOptions(node);
+
+  if (urls.length == 1) {
+    return urls.first;
+  }
+
+  print('Probing multiple options for $node backend:');
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/repositories/backend_urls.dart]: This code behaves differently locally vs staging vs prod, so better logging is required.

File: playground/frontend/playground_components/lib/src/repositories/backend_urls.dart [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+  final lastUrl = urls.removeLast();
+
+  for (final url in urls) {
+    try {
+      final client = GrpcExampleClient(url: url);
+      await client.getMetadata();
+      print('Using $url');
+      return url;
+    } catch (ex) {
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/repositories/backend_urls.dart]: `+`

**Comment by MakarkinSAkvelon** [comment]
LGTM

File: playground/frontend/playground_components/tools/generate_build_metadata.sh (hunk 14 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/tools/generate_build_metadata.sh]: `+`

File: playground/frontend/playground_components/lib/src/repositories/code_repository.dart (hunk 8 above)
Comment (alexeyinkin) [reply playground/frontend/playground_components/lib/src/repositories/code_repository.dart]: `+`

**Review by alexeyinkin — APPROVED** [review]

**Comment by alexeyinkin** [comment]
R: @damondouglas 

The only failing check is due to snippet renaming, it is addressed in
- https://github.com/apache/beam/pull/25033

Not merging it here to simplify separate PRs and reduce conflicts.

File: learning/tour-of-beam/backend/README.md [hunk 18]
```diff
@@ -39,6 +39,7 @@ To re-generate:
 $ go generate -x ./...
 ```
 
+> Note: this requires [`protoc`](https://grpc.io/docs/protoc-installation/) version `>= 3.15` and [`moq`](https://github.com/matryer/moq) tools to be installed
```
Comment (damondouglas) [inline learning/tour-of-beam/backend/README.md:42]: Does this generate the proto stubs?  I think we should either use buf which is already used for the Beam Playground here.  Or convert the Beam Playground to this.  Either way, both should be the same.

File: playground/backend/cmd/server/controller.go [hunk 19]
```diff
@@ -529,3 +529,31 @@ func (controller *playgroundController) GetSnippet(ctx context.Context, info *pb
...
+		}
+		return &timestamp
+	}()
+
+	buildCommitHash := func() *string {
+		if BuildCommitHash == "" {
+			return nil
+		}
+		return &BuildCommitHash
+	}()
```
Comment (damondouglas) [inline playground/backend/cmd/server/controller.go:549]: This unreadable code block is due to the use of the optional keyword [here](https://github.com/akvelon/beam/blob/get-metadata-endpoint-playground/playground/api/v1/api.proto#L323) [which is discouraged](https://cloud.google.com/apis/design/proto3). The optional keyword leads to a string* instead of just a string.

File: playground/api/v1/api.proto [hunk 20]
```diff
@@ -314,6 +314,17 @@ message GetSnippetResponse {
...
 
+// GetMetadataRequest represents request for runner metadata
+message GetMetadataRequest {}
+
+// GetMetadataResponse contains metadata about the runner
+message GetMetadataResponse {
+  string runner_sdk = 1;
+  optional string build_commit_hash = 2;
+  optional int64 build_commit_timestamp_seconds_since_epoch = 3;
+  optional string beam_sdk_version = 4;
```
Comment (damondouglas) [inline playground/api/v1/api.proto:325]: > Beginning in protobuf v3.14, primitive fields can distinguish between the default value and unset value by using the [optional keyword](https://cloud.google.com/apis/design/design_patterns.md#optional_primitive_fields), although this is generally discouraged.

See: https://cloud.google.com/apis/design/proto3

File: playground/backend/containers/git-functions.gradle [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+ext.getGitCommitTimestamp = () -> {
+   def stdout = new ByteArrayOutputStream()
+   exec {
+      executable('git')
+      args('show', '-s', '--format=%ct', 'HEAD')
+      standardOutput = stdout
+   }
+   return stdout.toString().trim()
+} as Closure<String>
```
Comment (damondouglas) [inline playground/backend/containers/git-functions.gradle:37]: Why is a gradle wrapper needed? Why not just run the command in the context where the value is needed?

File: playground/backend/containers/go/build.gradle [hunk 22]
```diff
@@ -83,19 +84,22 @@ task copyDockerfileDependencies(type: Copy) {
...
+   tags containerImageTags()
+   buildArgs(
+         ['BASE_IMAGE'   : project.rootProject.hasProperty(["base-image"]) ?
+               project.rootProject["base-image"] :
+               "golang:1.18-bullseye",
+          'SDK_TAG'      : project.rootProject.hasProperty(["sdk-tag"]) ?
+                project.rootProject["sdk-tag"] : project.rootProject.sdk_version,
+          'SDK_TAG_LOCAL': project.rootProject.sdk_version,
+          'GIT_COMMIT'   : getGitCommitHash(),
+          'GIT_TIMESTAMP': getGitCommitTimestamp()])
```
Comment (damondouglas) [inline playground/backend/containers/go/build.gradle:102]: The risk of wrapping more and more commands with gradle, the difficult it will be to troubleshoot in the future.

File: playground/backend/containers/java/Dockerfile [hunk 23]
```diff
@@ -55,6 +57,7 @@ FROM apache/beam_java8_sdk:$BEAM_VERSION
 ARG BEAM_VERSION
 ARG SPRING_VERSION=5.3.25
 ARG KAFKA_CLIENTS_VERSION=2.3.1
+ENV BEAM_VERSION=$BEAM_VERSION
```
Comment (damondouglas) [inline playground/backend/containers/java/Dockerfile:60]: We should specify environment variables outside the Docker image using ConfigMap

**Review by damondouglas — CHANGES_REQUESTED** [review]

File: playground/backend/containers/git-functions.gradle (hunk 21 above)
Comment (alexeyinkin) [reply playground/backend/containers/git-functions.gradle:37]: We need this in 6 separate locations:
- Building the frontend container.
- Building the router container.
- Building the 4 runners.

The 5 backends may be similar but frontend and backend are still two rather distinct users of this. This motivates to do this in a single location like this because duplicating `git show -s --format=%ct HEAD` and `git rev-parse HEAD` 2-6 times is more confusing.

Next, we cannot do this in Dockerfile because by the time it is built the directory is unaware of `.git`.

This leads to the requirement of having the values before the 6 containers' builds start. And their builds are triggered from Gradle. So we cannot shorten this to `command $(git rev-parse HEAD)`, we would still have the whole `exec { ... }` thing but in 12 locations.

File: playground/api/v1/api.proto (hunk 20 above)
Comment (alexeyinkin) [reply playground/api/v1/api.proto:325]: From the same link https://cloud.google.com/apis/design/design_patterns.md#optional_primitive_fields
> If not using optional would add complexity or ambiguity, then use optional primitive fields.

This is why I asked for `optional` here. This endpoint exists in both router and runners. The router is unaware of Beam and would have to return a magic value instead if omitting it. I think that omitting it on the protocol level is better.

There is indeed a bug in the current Dart implementation. It parses the raw data correctly, note the `null` in the router's response for Beam version:

image

But then the getter returns an empty string for it.

The issue to make those getters nullable is the most thumbed-up one in the repository, and members agree: https://github.com/google/protobuf.dart/issues/523

So I believe it will be done at some point. So if the support is almost there, I prefer to provision this right at the high level of the protocol to be able to improve the implementation soon after.

I found no other explanations of discouraging `optional` so I vote for using it.

There is another reason for going `optional`. On the frontend we have many enums coming from the backend: SDK, Complexity, etc. In protobuf they come as magic values like `SDK_UNDEFINED`. So we must maintain two versions of such enum classes: one with this value to be parsed from the protofub, and the other one is without it to be used in business logic. Converting between the two is complexity, so in the next phase of Playground development we would like to switch those to optional too. It would then be consistent with optional fields in this endpoint.

File: playground/backend/internal/environment/environment_service.go [hunk 24]
```diff
@@ -35,6 +35,7 @@ const (
 	serverIpKey                        = "SERVER_IP"
 	serverPortKey                      = "SERVER_PORT"
 	beamSdkKey                         = "BEAM_SDK"
+	beamVersionKey                = "BEAM_VERSION"
```
Comment (rshamunov) [inline playground/backend/internal/environment/environment_service.go:38]: Fix tabulation

File: playground/backend/containers/java/Dockerfile (hunk 23 above)
Comment (TSultanov) [reply playground/backend/containers/java/Dockerfile:60]: This is not a runtime configuration parameter in the strict sense. This variable holds the value of Beam SDK version with which the image was built. Storing this value outside of the container makes little sense to me.

File: playground/frontend/CONTRIBUTE.md [hunk 25]
```diff
@@ -66,25 +66,6 @@ contains decoupled playground page state and module providers like `OutputPlacem
 For quick start up, please take a look
 to [this guide](https://docs.flutter.dev/development/data-and-backend/state-mgmt/simple)
 
-## Configuration
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply playground/frontend/CONTRIBUTE.md:69]: `+`

File: playground/api/v1/api.proto (hunk 20 above)
Comment (TSultanov) [reply playground/api/v1/api.proto]: Removed the `optional` fields

File: learning/tour-of-beam/backend/README.md (hunk 18 above)
Comment (TSultanov) [reply learning/tour-of-beam/backend/README.md:42]: Created task #25775 to resolve this in a separate PR

**Review by damondouglas — CHANGES_REQUESTED** [review]
I ran through the entire build/deploy solution documented at https://github.com/akvelon/beam/blob/get-metadata-endpoint-playground/playground/terraform/README.md and I'm seeing CORS errors from the domain I specified `-Pdns-name=` in the documentation.  I was unable to test through the installed application whether this PR broke anything.  The images built successfully and the k8s manifests installed correctly.

File: playground/backend/containers/go/build.gradle (hunk 22 above)
Comment (damccorm) [reply playground/backend/containers/go/build.gradle:102]: IMO this is ok in order to get the reusability
